File: cpp_src/puppy_core/safety_manager.cpp
// Safety Manager 实现
#include "puppy_core/safety_manager.h"

#include <algorithm>

namespace puppy_core {

SafetyManagerNode::SafetyManagerNode() : rclcpp::Node("safety_manager") {
  // P1-2: 安全参数从 declare_parameter 读取（可由 safety.yaml 覆盖）
  this->declare_parameter("cmd_vel_timeout", 1.0);
  this->declare_parameter("low_battery_threshold", 0.20);
  this->declare_parameter("critical_battery_threshold", 0.10);
  this->declare_parameter("max_linear_x", 0.3);
  this->declare_parameter("max_angular_z", 1.2);
  this->declare_parameter("motors_enabled_on_start", false);

  cmd_vel_timeout_ = this->get_parameter("cmd_vel_timeout").as_double();
  max_linear_x_ = this->get_parameter("max_linear_x").as_double();
  max_angular_z_ = this->get_parameter("max_angular_z").as_double();
  low_battery_threshold_ =
      this->get_parameter("low_battery_threshold").as_double();
  critical_battery_threshold_ =
      this->get_parameter("critical_battery_threshold").as_double();
  last_cmd_vel_time_ = this->now();
  // Default: disabled (gaijin2.md 14.6)
  motors_enabled_ = this->get_parameter("motors_enabled_on_start").as_bool();

  // Publishers
  safe_cmd_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
      "/cmd_vel_safe", 10);
  motor_enable_pub_ = this->create_publisher<std_msgs::msg::Bool>(
      "/robot/motion_enable", 10);

  // Subscribers
  cmd_vel_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
      "/cmd_vel", 10,
      std::bind(&SafetyManagerNode::onCmdVel, this, std::placeholders::_1));
  battery_sub_ =
      this->create_subscription<puppy_interfaces::msg::BatteryStatus>(
          "/battery_status", 10,
          std::bind(&SafetyManagerNode::onBattery, this,
                    std::placeholders::_1));
  fall_sub_ = this->create_subscription<puppy_interfaces::msg::FallEvent>(
      "/fall/event", 10,
      std::bind(&SafetyManagerNode::onFall, this, std::placeholders::_1));
  motor_request_sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/robot/motors_enable_request", 10,
      std::bind(&SafetyManagerNode::onMotorRequest, this,
                std::placeholders::_1));

  // Safety check timer (20Hz)
  timer_ = this->create_wall_timer(
      std::chrono::milliseconds(50),
      std::bind(&SafetyManagerNode::safetyCheck, this));

  RCLCPP_INFO(this->get_logger(),
              "Safety manager started (motors DISABLED by default)");
}

void SafetyManagerNode::onCmdVel(
    const geometry_msgs::msg::Twist::SharedPtr msg) {
  // Forward cmd_vel with safety checks.
  last_cmd_vel_time_ = this->now();
  if (!motors_enabled_) {
    // Motors disabled - send zero velocity
    geometry_msgs::msg::Twist zero;
    safe_cmd_pub_->publish(zero);
    return;
  }
  // Apply safety limits
  geometry_msgs::msg::Twist safe_msg;
  safe_msg.linear.x = std::max(-max_linear_x_,
                               std::min(max_linear_x_, msg->linear.x));
  safe_msg.angular.z = std::max(-max_angular_z_,
                                std::min(max_angular_z_, msg->angular.z));
  safe_cmd_pub_->publish(safe_msg);
}

void SafetyManagerNode::onBattery(
    const puppy_interfaces::msg::BatteryStatus::SharedPtr msg) {
  // Monitor battery for safety thresholds.
  if (msg->critical_battery) {
    RCLCPP_ERROR(this->get_logger(), "Critical battery! Entering SAFE_STOP");
    motors_enabled_ = false;
    publishMotorEnable();
  }
}

void SafetyManagerNode::onFall(
    const puppy_interfaces::msg::FallEvent::SharedPtr msg) {
  // Fall detection - immediately disable motors.
  if (msg->detected) {
    RCLCPP_ERROR(this->get_logger(),
                 "Fall detected (%s)! Stopping motors",
                 msg->direction.c_str());
    motors_enabled_ = false;
    publishMotorEnable();
  }
}

void SafetyManagerNode::onMotorRequest(
    const std_msgs::msg::Bool::SharedPtr msg) {
  // Handle motor enable/disable requests.
  motors_enabled_ = msg->data;
  RCLCPP_INFO(this->get_logger(), "Motors %s",
              msg->data ? "enabled" : "disabled");
  publishMotorEnable();
}

void SafetyManagerNode::safetyCheck() {
  // Periodic safety check - command timeout.
  if (motors_enabled_) {
    double elapsed = (this->now() - last_cmd_vel_time_).seconds();
    if (elapsed > cmd_vel_timeout_) {
      // Command timeout - publish zero velocity
      geometry_msgs::msg::Twist zero;
      safe_cmd_pub_->publish(zero);
    }
  }
}

void SafetyManagerNode::publishMotorEnable() {
  std_msgs::msg::Bool msg;
  msg.data = motors_enabled_;
  motor_enable_pub_->publish(msg);
}

}  // namespace puppy_core

File: src/puppy_core/puppy_core/safety_manager.py
"""Safety Manager: monitors safety conditions and triggers emergency response.

Implements the safety requirements (gaijin2.md section 14):
  1. Command timeout auto-stop
  2. Mode switch clears motion commands
  3. Fall detection stops navigation
  4. Low battery prevents patrol
  5. Driver error enters SAFE_STOP
  6. Motors disabled by default on startup
"""
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import Bool
from puppy_interfaces.msg import BatteryStatus, FallEvent


class SafetyManagerNode(Node):
    """Monitors safety conditions and enforces emergency responses."""

    def __init__(self):
        super().__init__('safety_manager')

        # P1-2: 安全参数从 declare_parameter 读取（可由 safety.yaml 覆盖）
        self.declare_parameter('cmd_vel_timeout', 1.0)
        self.declare_parameter('low_battery_threshold', 0.20)
        self.declare_parameter('critical_battery_threshold', 0.10)
        self.declare_parameter('max_linear_x', 0.3)
        self.declare_parameter('max_angular_z', 1.2)
        self.declare_parameter('motors_enabled_on_start', False)

        # Safety state
        self.cmd_vel_timeout = self.get_parameter('cmd_vel_timeout').value
        self.max_linear_x = self.get_parameter('max_linear_x').value
        self.max_angular_z = self.get_parameter('max_angular_z').value
        self.low_battery_threshold = self.get_parameter('low_battery_threshold').value
        self.critical_battery_threshold = self.get_parameter('critical_battery_threshold').value
        self.last_cmd_vel_time = self.get_clock().now()
        # Default: disabled (gaijin2.md 14.6)
        self.motors_enabled = self.get_parameter('motors_enabled_on_start').value

        # Publishers
        self.safe_cmd_pub = self.create_publisher(Twist, '/cmd_vel_safe', 10)
        self.motor_enable_pub = self.create_publisher(Bool, '/robot/motion_enable', 10)

        # Subscribers
        self.create_subscription(Twist, '/cmd_vel', self._on_cmd_vel, 10)
        self.create_subscription(BatteryStatus, '/battery_status', self._on_battery, 10)
        self.create_subscription(FallEvent, '/fall/event', self._on_fall, 10)
        self.create_subscription(Bool, '/robot/motors_enable_request', self._on_motor_request, 10)
        # 看门狗急停: 传感器失效/断链/心跳超时由 watchdog 判定后经此话题下达。
        # 此前仲裁者未订阅它, 导致看门狗急停无法作用到执行器 (真实缺口)。
        self.create_subscription(Bool, '/watchdog/emergency_stop',
                                 self._on_emergency_stop, 10)

        # Safety check timer (20Hz)
        self.create_timer(0.05, self._safety_check)

        self.get_logger().info('Safety manager started (motors DISABLED by default)')

    def _on_cmd_vel(self, msg: Twist):
        """Forward cmd_vel with safety checks."""
        self.last_cmd_vel_time = self.get_clock().now()
        if not self.motors_enabled:
            # Motors disabled - send zero velocity
            self.safe_cmd_pub.publish(Twist())
            return
        # Apply safety limits
        safe_msg = Twist()
        safe_msg.linear.x = max(-self.max_linear_x, min(self.max_linear_x, msg.linear.x))
        safe_msg.angular.z = max(-self.max_angular_z, min(self.max_angular_z, msg.angular.z))
        self.safe_cmd_pub.publish(safe_msg)

    def _on_emergency_stop(self, msg: Bool):
        """看门狗急停: 立即断电机使能并输出零速度 (§4 P0-4 急停必须零速度)。"""
        if msg.data:
            self.get_logger().error('Watchdog EMERGENCY STOP -> SAFE_STOP')
            self.motors_enabled = False
            self.safe_cmd_pub.publish(Twist())
            self._publish_motor_enable()

    def _on_battery(self, msg: BatteryStatus):
        """Monitor battery for safety thresholds."""
        if msg.critical_battery:
            self.get_logger().error('Critical battery! Entering SAFE_STOP')
            self.motors_enabled = False
            self._publish_motor_enable()

    def _on_fall(self, msg: FallEvent):
        """Fall detection - immediately disable motors."""
        if msg.detected:
            self.get_logger().error(f'Fall detected ({msg.direction})! Stopping motors')
            self.motors_enabled = False
            self._publish_motor_enable()

    def _on_motor_request(self, msg: Bool):
        """Handle motor enable/disable requests."""
        self.motors_enabled = msg.data
        self.get_logger().info(f'Motors {"enabled" if msg.data else "disabled"}')
        self._publish_motor_enable()

    def _safety_check(self):
        """Periodic safety check - command timeout."""
        if self.motors_enabled:
            elapsed = (self.get_clock().now() - self.last_cmd_vel_time).nanoseconds / 1e9
            if elapsed > self.cmd_vel_timeout:
                # Command timeout - publish zero velocity
                self.safe_cmd_pub.publish(Twist())

    def _publish_motor_enable(self):
        msg = Bool()
        msg.data = self.motors_enabled
        self.motor_enable_pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = SafetyManagerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()
